test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -1620,6 +1693,429 @@ TEST_F(ClusterManagerImplTest, OriginalDstInitialization) {
   factory_.tls_.shutdownThread();
 }
 
+TEST_F(ClusterManagerImplTest, MergedUpdates) {
+  createWithLocalClusterUpdate();
+
+  // Ensure we see the right set of added/removed hosts on every call.
+  EXPECT_CALL(local_cluster_update_, post(_, _, _))
+      .WillRepeatedly(Invoke([](uint32_t priority, const HostVector& hosts_added,
+                                const HostVector& hosts_removed) -> void {
+        static int call = 0;
```

Comment:
This can be structured in a more readable way by a sequence of `.WillOnce(doFoo).WillOnce(doBar)` etc. Rather than maintaining an internal state machine.